core/client/src/main/java/alluxio/client/file/UnderFileSystemUtils.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.client.file;
+
+import alluxio.AlluxioURI;
+import alluxio.collections.Pair;
+import alluxio.exception.AlluxioException;
+import alluxio.security.authorization.Permission;
+import alluxio.underfs.UnderFileSystem;
+import alluxio.underfs.options.MkdirsOptions;
+
+import java.io.IOException;
+import java.util.Stack;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Utility functions for working with {@link alluxio.underfs.UnderFileSystem}.
+ *
+ */
+@ThreadSafe
+public final class UnderFileSystemUtils {
+
+  /**
+   * Creates parent directories for path with correct permissions if required.
+   *
+   * @param alluxioPath Alluxio path
+   * @param ufsPath path in the under file system
+   * @param fs file system master client
+   * @param ufs the under file system
+   * @throws AlluxioException if an unexpected Alluxio exception is thrown
+   * @throws IOException if folder creation fails
+   */
+  public static void prepareFilePath(AlluxioURI alluxioPath, String ufsPath,
+      FileSystem fs, UnderFileSystem ufs) throws AlluxioException, IOException {
+    AlluxioURI dstPath = new AlluxioURI(ufsPath);
+    String parentPath = dstPath.getParent().getPath();
+    // creates the parent folder if it does not exist
+    if (!ufs.isDirectory(parentPath)) {
+      // Create ancestor directories from top to the bottom. We cannot use recursive create parents
+      // here because the permission for the ancestors can be different.
+      Stack<Pair<String, MkdirsOptions>> ufsDirsToMakeWithOptions = new Stack<>();
+      AlluxioURI curAlluxioPath = alluxioPath.getParent();
+      AlluxioURI curUfsPath = dstPath.getParent();
+      // Stop at the Alluxio root because the mapped directory of Alluxio root in UFS may not exist.
```

Comment:
True. The comment has gone stale.

If the mount point the current path is under is not deleted after mounting, the loop will stop at the mount point as ufs.isDirectory will return true. But you're right to handle the case where it might be, we need to stop at mount point.